src/main/java/org/reactome/addlinks/fileprocessors/ensembl/EnsemblBiomartFileProcessor.java

```diff
@@ -0,0 +1,115 @@
+    {
+        super(processorName);
+    }
+
+    // TODO: Add unit tests
+    /**
+     * This class builds a large, rather unwieldy two-layer mapping structure of identifiers corresponding to a specific species.
+     * The first level has 4 types of keys for each species. These keys are 'uniprotToProteins', 'proteinToGenes',
+     * 'proteinToTranscripts', and 'transcriptToMicroarrays'. These describe the type of mapping that this
+     * key is to. For example, the key 'hsapiens_uniprotToProteins' would access a Map of Human UniProt protein identifiers
+     * to a List of Ensembl protein identifiers. Transcript denotes Ensembl transcript identifiers, Gene is Ensembl gene identifiers
+     * and Microarray denotes microarray probe identifiers.
+     * @return Map<String, Map<String, List<String>>>> mappings, the two-layer mapping structure of species-(identifier-[identifiers]).
+     * @throws IOException Can be thrown if file does not exist.
+     */
+    @Override
+    public Map<String, Map<String, List<String>>> getIdMappingsFromFile() throws IOException {
+
+        Map<String, Map<String, List<String>>> mappings = new HashMap<>();
+        // Iterate through each biomart file in the addlinks directory
+        File biomartDir = new File(this.pathToFile.toString());
+        for (File biomartFile : biomartDir.listFiles()) {
+            // Get species name from filename
+            String biomartFileSpecies = biomartFile.getName().split("_")[0];
+            // Mappings that will be generated from each file for the species.
+            Map<String, List<String>> proteinToTranscripts = new HashMap<>();
+            Map<String, List<String>> proteinToGenes = new HashMap<>();
+            Map<String, List<String>> transcriptToMicroarrays = new HashMap<>();
+            Map<String, List<String>> uniprotToProteins = new HashMap<>();
+            // Read file and iterate through each line.
+            BufferedReader br = new BufferedReader(new FileReader(biomartFile));
+            String biomartLine;
+            while ((biomartLine = br.readLine()) !=null) {
+                // Split each tab-separated line. Each line has four values. The first three are always
+                // Ensembl Gene (ENSG), Transcript (ENST) and Protein (ENSP), in that order.
+                // The 4th can be a UniProt or microarray identifier, depending on the file being read.
+                // All 4 values may not be in each line.
+                List<String> tabSplit = Arrays.asList(biomartLine.split("\t"));
+                // Processing of UniProt mapping files.
+                // UniProt identifier mapping files are used to fully populate
+                // the 'uniprotToProteins' and 'proteinToGenes' mappings.
+                if (biomartFile.getName().contains("uniprot")) {
+                    // 'uniprotToProteins' mapping requires the 3rd (protein) and 4th (uniprot)values in the line.
```

Comment:
Nitpick - space needed between "(uniprot)" and "values"?

Reply:
> Nitpick - space needed between "(uniprot)" and "values"?

:roll_eyes: